File: stats-scraper/src/services/mockData.ts
import type { DataColumnSpec, DatasetSchema, GeneratedDataset, InsightCandidate } from '../domain/types'

type MockDataOptions = {
  seed?: number
  rowCount?: number
}

const createSeededRandom = (seed: number): (() => number) => {
  let state = (seed >>> 0) || 1

  return () => {
    state = (state * 1664525 + 1013904223) >>> 0
    return state / 4294967296
  }
}

const clamp = (value: number, min?: number, max?: number): number => {
  let result = value
  if (min !== undefined && result < min) result = min
  if (max !== undefined && result > max) result = max
  return result
}

const generateCategory = (spec: DataColumnSpec, random: () => number): unknown => {
  const categories = spec.categories ?? ['A', 'B', 'C']
  return categories[Math.floor(random() * categories.length)]
}

const generateNormal = (spec: DataColumnSpec, random: () => number): number => {
  const mean = spec.mean ?? 0
  const stddev = spec.stddev ?? 1
  const u1 = Math.max(random(), 1e-10)
  const u2 = random()
  const z = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2)
  return clamp(Math.round((mean + z * stddev) * 100) / 100, spec.min, spec.max)
}

const generateUniform = (spec: DataColumnSpec, random: () => number): number => {
  const min = spec.min ?? 0
  const max = spec.max ?? 100
  return clamp(Math.round((min + random() * (max - min)) * 100) / 100, spec.min, spec.max)
}

const generateLinear = (spec: DataColumnSpec, index: number): number => {
  const start = spec.start ?? 0
  const end = spec.end ?? 100
  const step = spec.step ?? (end - start) / 200
  return Math.round((start + index * step) * 100) / 100
}

const generateConstant = (spec: DataColumnSpec): unknown => {
  return spec.value ?? 0
}

const generateValue = (spec: DataColumnSpec, index: number, random: () => number): unknown => {
  switch (spec.generator) {
    case 'category':
      return generateCategory(spec, random)
    case 'normal':
      return generateNormal(spec, random)
    case 'uniform':
      return generateUniform(spec, random)
    case 'linear':
      return generateLinear(spec, index)
    case 'constant':
      return generateConstant(spec)
    default:
      return null
  }
}

export const generateMockDataset = (
  _schema: DatasetSchema,
  insight: InsightCandidate,
  opts: MockDataOptions = {}
): GeneratedDataset => {
  const seed = opts.seed ?? 1337
  const profile = insight.dataProfile
  const rowCount = opts.rowCount ?? profile?.rowCount ?? 200
  const random = createSeededRandom(seed)
  const columns = profile?.columns ?? []
  const columnNames = columns.map((col) => col.name)

  const rows = Array.from({ length: rowCount }, (_, rowIndex) => {
    const row: Record<string, unknown> = {}

    for (const col of columns) {
      row[col.name] = generateValue(col, rowIndex, random)
    }

    return row
  })

  return {
    id: `dataset-${insight.id}-${seed}`,
    name: `${insight.title} sample`,
    columns: columnNames.length > 0 ? columnNames : ['x', 'y'],
    rows
  }
}

File: backend/mock_data.py
"""Mock data generator — Python port of frontend mockData.ts.

Generates deterministic mock data from an insight's dataProfile.
Uses the same Linear Congruential Generator algorithm as the frontend.
"""

from typing import Any

MOCK_ROW_COUNT = 200


def _create_seeded_random(seed: int):
    state = seed & 0xFFFFFFFF or 1

    def random():
        nonlocal state
        state = (state * 1664525 + 1013904223) & 0xFFFFFFFF
        return state / 4294967296

    return random


def _clamp(value: float, min_val: Any, max_val: Any) -> float:
    result = value
    if min_val is not None and result < min_val:
        result = min_val
    if max_val is not None and result > max_val:
        result = max_val
    return result


def _generate_category(spec: dict, random) -> Any:
    categories = spec.get("categories") or ["A", "B", "C"]
    return categories[int(random() * len(categories))]


def _generate_normal(spec: dict, random) -> float:
    mean = spec.get("mean") or 0
    stddev = spec.get("stddev") or 1
    u1 = max(random(), 1e-10)
    u2 = random()
    z = math.sqrt(-2 * math.log(u1)) * math.cos(2 * math.pi * u2)
    return round(_clamp(
        round((mean + z * stddev) * 100) / 100,
        spec.get("min"), spec.get("max")
    ), 2)


def _generate_uniform(spec: dict, random) -> float:
    min_val = spec.get("min") or 0
    max_val = spec.get("max") or 100
    return round(_clamp(
        round((min_val + random() * (max_val - min_val)) * 100) / 100,
        spec.get("min"), spec.get("max")
    ), 2)


def _generate_linear(spec: dict, index: int) -> float:
    start = spec.get("start") or 0
    end = spec.get("end") or 100
    step = spec.get("step") or ((end - start) / 200 if end != start else 1)
    return round(start + index * step, 2)


def _generate_constant(spec: dict) -> Any:
    return spec.get("value") or 0


def _generate_value(spec: dict, index: int, random) -> Any:
    generator = spec.get("generator", "uniform")
    if generator == "category":
        return _generate_category(spec, random)
    elif generator == "normal":
        return _generate_normal(spec, random)
    elif generator == "uniform":
        return _generate_uniform(spec, random)
    elif generator == "linear":
        return _generate_linear(spec, index)
    elif generator == "constant":
        return _generate_constant(spec)
    return None


def _json_path_to_key(json_path: str) -> str:
    return json_path.lstrip("$.")


def generate_mock_rows(insight: dict, seed: int = 1337) -> list[dict]:
    """Generate mock data rows from an insight's dataProfile."""
    profile = insight.get("dataProfile")
    if not profile:
        return []

    columns = profile.get("columns", [])
    if not columns:
        return []

    random = _create_seeded_random(seed)
    rows = []

    for row_index in range(MOCK_ROW_COUNT):
        row = {}
        for col in columns:
            key = _json_path_to_key(col.get("name", ""))
            row[key] = _generate_value(col, row_index, random)
        rows.append(row)

    return rows
